### PR TITLE
porcelain.checkout

### PR DESCRIPTION
**Pull request by DedSecer**
Improve from [ pythonista stash's git cli](https://github.com/ywangd/stash/blob/master/bin/git.py)
This seemed to work in my simple testing, but I am not sure about if it actually work as `git checkout [branch]`

### CODE REVIEW

File: dulwich/porcelain.py [hunk 1]
```diff
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+                print('file not in index.', path.decode())
+            return
+
+        try:
+            index_entry = list(index[path])
+        except KeyError:
+            # if index_entry doesnt exist, this file was being removed. readd it
+            # update index entry stats to reflect commit
+            index_entry = [(0, 0), (0,0), 15, 0, 0, 1000, 1000, 0, tree_entry[1], 0, 0]
+        index_entry[0] = (repo[b'HEAD'].commit_time, 0)  #ctime
```
Comment (jelmer) [inline dulwich/porcelain.py:1785]: index_entry is a namedtuple nowadays, so you should be able to use index_entry.ctime here - which is slightly more readable.

File: dulwich/porcelain.py [hunk 2]
```diff
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+            except KeyError:
+                print('file not in index.', path.decode())
+            return
+
+        try:
+            index_entry = list(index[path])
+        except KeyError:
+            # if index_entry doesnt exist, this file was being removed. readd it
+            # update index entry stats to reflect commit
+            index_entry = [(0, 0), (0,0), 15, 0, 0, 1000, 1000, 0, tree_entry[1], 0, 0]
```
Comment (jelmer) [inline dulwich/porcelain.py:1784]: This is harcoding several values that don't necessarily make sense, e.g. uid/gid.

File: dulwich/porcelain.py [hunk 3]
```diff
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+        tree_id = repo[b'HEAD']._tree
+        try:
+            tree_entry = repo[tree_id].lookup_path(lambda x: repo[x], path)
+        except KeyError:
+            # if tree_entry didnt exist, this file was being added, so remove index entry
+            try:
+                del index[path]
+                index.write()
+            except KeyError:
+                print('file not in index.', path.decode())
```
Comment (jelmer) [inline dulwich/porcelain.py:1776]: this should probably raise an exception?

File: dulwich/porcelain.py [hunk 4]
```diff
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+            par_dir_tree_id = get_entry(repo, tree, os.path.split(file_path)[0])[1]
+            tree = parse_tree(repo, treeish=par_dir_tree_id)
+            tree_id = tree[os.path.split(file_path)[1].encode()]
+            return tree_id
+        else:
+            try:
+                return tree[file_path.encode()]
+            except KeyError:
+                # untracked file not in tree.
+                print('file not in tree')
```
Comment (jelmer) [inline dulwich/porcelain.py:1843]: this should probably raise an exception rather than write to standard out (which may not be visible). Another option would be for the function to return the list of paths for files it actually changed.

**Review by jelmer — CHANGES_REQUESTED** [review]
Thanks for the PR. A couple of high-level comments:

* Please add unit tests for any functionality you add (in this case, in dulwich/tests/test_porcelain.py)
* Dulwich uses a particular style for docstrings (Google style docstrings), this means e.g. double quotes rather than single quotes

File: dulwich/porcelain.py (hunk 2 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: Where can I get the other value in index_entry such as uid, gid etc. ? And it is necessary to set them?

File: dulwich/porcelain.py (hunk 4 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1842]: I have considered about it. But I am confusing about which type of Error to raise.

File: dulwich/porcelain.py [hunk 5]
```diff
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+            # update index entry stats to reflect commit
+            index_entry = [(0, 0), (0,0), 15, 0, 0, 1000, 1000, 0, tree_entry[1], 0, 0]
+        index_entry[0] = (repo[b'HEAD'].commit_time, 0)  #ctime
+        index_entry[1] = (repo[b'HEAD'].commit_time, 0)  #mtime
+        index_entry[4] = tree_entry[0]  #mode
+        index_entry[7] = len(repo[tree_entry[1]].data)  #size
+        index_entry[8] = tree_entry[1]  #sha
+
+        index[path] = index_entry
+        index.write()
```
Comment (jelmer) [inline dulwich/porcelain.py:1791]: You probably want to write the index outside of the loop, rather than once for every entry.

File: dulwich/porcelain.py (hunk 4 above)
Comment (jelmer) [reply dulwich/porcelain.py:1842]: Either KeyError or FileNotFoundError seem like they could be appropriate, or we could define a new exception type.

File: dulwich/porcelain.py (hunk 5 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: OK.That's my mistake.

File: dulwich/porcelain.py (hunk 3 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: It will raise KeyError in the new commit

File: dulwich/porcelain.py (hunk 4 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: OK,It will raise KeyError in the new commit.

File: dulwich/porcelain.py (hunk 2 above)
Comment (jelmer) [reply dulwich/porcelain.py]: We could set them to the current users' identity, or alternatively just to 0 (which may be easiest).

**Comment by DedSecer** [comment]
Do you think if it will be better to move the function `unstage` from `porcelain` under `dulwich.repo.Repo`?

**Comment by jelmer** [comment]
> Do you think if it will be better to move the function `unstage` from `porcelain` under `dulwich.repo.Repo`?

We should keep an eye on the number of methods on Repo, since it's already quite large. But adding Repo.unstage makes sense I think.

**Comment by DedSecer** [comment]
I have added unittest for `stage`, `reset_file` and `checkout` now.

File: dulwich/errors.py [hunk 6]
```diff
@@ -84,6 +84,10 @@ class NotBlobError(WrongObjectException):
     type_name = "blob"
 
 
+class CheckoutError(Exception):
```
Comment (jelmer) [inline dulwich/errors.py:87]: Can you give this a less generic name? There are all kinds of errors that can occur during checkout.

File: dulwich/porcelain.py [hunk 7]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
             r.refs.set_symbolic_ref(b"HEAD", to_set)
 
 
+def reset_file(repo, file_path: str , target: bytes = b'HEAD'):
+    """
+    reset the file to specific commit or branch
```
Comment (jelmer) [inline dulwich/porcelain.py:1760]: Per google doc string style and the rest of Dulwich, please Put the summary immediately after the triple quotes, capitalize and end with a trailing period.

File: dulwich/porcelain.py [hunk 8]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+    Args:
+        repo: dulwich Repo object
```
Comment (jelmer) [inline dulwich/porcelain.py:1762]: For consistency, just two spaces.

File: dulwich/porcelain.py [hunk 9]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+        file_path: file to reset, relative to the repository path
+        target: branch or commit or b'HEAD' to reset
+    """
+    if target in branch_list(repo):
```
Comment (jelmer) [inline dulwich/porcelain.py:1766]: you probably want just parse_tree() here rather than duplicating the logic (see elsewhere in dulwich/porcelain.py)

File: dulwich/porcelain.py [hunk 10]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+        sha = repo.refs[b'refs/heads/' + target]
+    elif target == b'HEAD':
+        sha = repo.head()
+    else:
+        sha = target
+    tree = parse_tree(repo, treeish=sha)
+
+    def get_entry(repo, tree, file_path: str):
```
Comment (jelmer) [inline dulwich/porcelain.py:1774]: get_entry could be replaced with tree.lookup_path(repo.object_store.__getitem__, file_path)

File: dulwich/porcelain.py [hunk 11]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+    file_entry = get_entry(repo, tree, file_path)
+    
+    full_path = os.path.join(repo.path, file_path)
+    blob = repo.object_store[file_entry[1]]
+    mode = file_entry[0]
+    build_file_from_blob(blob, mode, full_path.encode())
+
+
+def checkout(repo, branch: bytes, force: bool = False):
+    """
```
Comment (jelmer) [inline dulwich/porcelain.py:1796]: See other comment about docstrings.

File: dulwich/porcelain.py [hunk 12]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+    # unstage all files in the index
+    tracked_changes = []
+    for file in get_tree_changes(repo).values():
+        tracked_changes += file
+    repo.unstage(tracked_changes)
+
+    # reset all unstaged file to target
+    normalizer = repo.get_blob_normalizer()
+    filter_callback = normalizer.checkin_normalize
+    unstaged_files = list(get_unstaged_changes(repo.open_index(), repo.path, filter_callback))
```
Comment (jelmer) [inline dulwich/porcelain.py:1818]: why list(), couldn't you just iterate through the entries of get_unstaged_changes?

File: dulwich/repo.py [hunk 13]
```diff
@@ -1305,6 +1305,59 @@ def stage(self, fs_paths):
                     index[tree_path] = index_entry_from_stat(st, blob.id, 0)
         index.write()
 
+    def unstage(self, relpaths: List[bytes] = []):
+        """
+        unstage specific file in the index
```
Comment (jelmer) [inline dulwich/repo.py:1310]: See earlier comment about docstrings.

File: dulwich/tests/test_porcelain.py [hunk 14]
```diff
@@ -1317,6 +1320,222 @@ def test_hard_commit(self):
         self.assertEqual([], changes)
 
 
+class UnstageTests(PorcelainTestCase):
```
Comment (jelmer) [inline dulwich/tests/test_porcelain.py:1323]: Since unstage lives on Repo, please add these tests to dulwich/tests/test_repo.py

File: dulwich/porcelain.py [hunk 15]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+    """
+    switch branches or restore working tree files
+    Args:
+        repo: dulwich Repo object
+        branch: branch name to checkout
+    """
+
+    # check repo status
+    if not force:
+        if status(repo) != GitStatus(staged={'add': [], 'delete': [], 'modify': []}, unstaged=[], untracked=[]):
```
Comment (jelmer) [inline dulwich/porcelain.py:1805]: What about ignored files? They could be versioned in the target.

File: dulwich/porcelain.py [hunk 16]
```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+    for file in get_tree_changes(repo).values():
+        tracked_changes += file
+    repo.unstage(tracked_changes)
+
+    # reset all unstaged file to target
+    normalizer = repo.get_blob_normalizer()
+    filter_callback = normalizer.checkin_normalize
+    unstaged_files = list(get_unstaged_changes(repo.open_index(), repo.path, filter_callback))
+    for file in unstaged_files:
+        reset_file(repo, file.decode(), b'HEAD')
```
Comment (jelmer) [inline dulwich/porcelain.py:1820]: wouldn't this need to update the index as well?

File: dulwich/porcelain.py (hunk 15 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: It seem that ignored files will not appear in the GitSatus, and `git checkout` will not modify the ignored files

Comment (jelmer) [reply dulwich/porcelain.py]: The branch you're changing to can contain files that were ignored (and not versioned) in the original branch you were on.

Comment (DedSecer) [reply dulwich/porcelain.py]: Does your meaning like this?
```
$ git init
$ # add new thing to file `foo`
$ git add foo; git commit -m 'empty'
$ git branch bran
$ # add new thing to file `bar`
$ # add file `bar` to .gitignore
$ git add .gitignore; git commit -m 'empty1'
```
Here's the output of `git status` in branch `master`:
```
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        bar

nothing added to commit but untracked files present (use "git add" to track)
```

And the output of `git status` in branch `bran`:
```
On branch bran
nothing to commit, working tree clean
```

I think both `git checkout` and  `porcelain.checkout` will not do any thing to the file `bar` after the first commit

File: dulwich/errors.py (hunk 6 above)
Comment (DedSecer) [reply dulwich/errors.py]: raising DirNotCleanError in the new commit

File: dulwich/porcelain.py (hunk 10 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1774]: I have replace `get_entry` with `tree.lookup_path(repo.object_store.__getitem__, file_path)`, 
but the test rasing error in the `dulwich.tests.test_porcelain.ResetFileTests`:
```
> python -m unittest dulwich.tests.test_porcelain.ResetFileTests
...E
======================================================================
ERROR: test_resetfile_with_dir (dulwich.tests.test_porcelain.ResetFileTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\Repo\dulwich\dulwich\tests\test_porcelain.py", line 1497, in test_resetfile_with_dir
  File "D:\Repo\dulwich\dulwich\porcelain.py", line 1781, in reset_file
    file_entry = tree.lookup_path(repo.object_store.__getitem__, file_path.encode())
  File "D:\Repo\dulwich\dulwich\objects.py", line 1191, in lookup_path
    mode, sha = obj[p]
  File "D:\Repo\dulwich\dulwich\objects.py", line 1058, in __getitem__
    return self._entries[name]
KeyError: b'new_dir\\foo'
```
I think we should change `parts = path.split(b"/")` with `parts = os.path.split(path)` in the [objects.py, line 1182](https://github.com/dulwich/dulwich/blob/80bffaca1dfe59a5ba57a045f3ce7122227a7647/dulwich/objects.py#L1182)

File: dulwich/porcelain.py (hunk 9 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1765]: I have remove other Conditionals except
```
if target in branch_list(repo):
        target = b'refs/heads/' + target
```
I think we should keep this to match branch name

Comment (jelmer) [reply dulwich/porcelain.py:1765]: If we'd like to support branch names here (and I agree that would be a good idea), then we should fix that in parse_tree rather than here.

File: dulwich/porcelain.py [hunk 17]
```diff
@@ -1753,6 +1755,61 @@ def update_head(repo, target, detached=False, new_branch=None):
             r.refs.set_symbolic_ref(b"HEAD", to_set)
 
 
+def reset_file(repo, file_path: str , target: bytes = b'HEAD'):
```
Comment (jelmer) [inline dulwich/porcelain.py:1758]: extra space after "str"

**Comment by jelmer** [comment]
Thanks for your patience; I think we're close to being able to merge this now.

File: dulwich/porcelain.py (hunk 17 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1758]: If I remove extra space after "str", the flake8 will output: `missing whitespace after ':'flake8(E231)`

File: dulwich/porcelain.py (hunk 9 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: Update parser_tree in the new commit

File: dulwich/porcelain.py (hunk 16 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1826]: reset_file will update index now

Comment (DedSecer) [reply dulwich/porcelain.py:1829]: Where we should update index?After reset_file or at the end of checkout?

**Comment by DedSecer** [comment]
@jelmer you can have a review now.

File: dulwich/porcelain.py [hunk 18]
```diff
@@ -1753,6 +1756,91 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+        
+        index = repo.open_index()
+        normalizer = repo.get_blob_normalizer()
+        filter_callback = normalizer.checkin_normalize
+        unstaged_changes = list(get_unstaged_changes(index, repo.path, filter_callback))
+        for file in unstaged_changes:
+            if file in repo.open_index():
+                raise DirNotCleanError('trying to checkout when working directory not clean')
+ 
+    current_file_set = set(parse_tree(repo, repo.head()))
```
Comment (jelmer) [inline dulwich/porcelain.py:1813]: this will only return files in the root; how would this work for files under directories?

File: dulwich/porcelain.py (hunk 16 above)
Comment (jelmer) [reply dulwich/porcelain.py:1836]: After the checkout, otherwise this will be very slow for larger repositories.

File: dulwich/objectspec.py [hunk 19]
```diff
@@ -54,6 +54,8 @@ def parse_tree(repo, treeish):
       KeyError: If the object can not be found
     """
     treeish = to_bytes(treeish)
+    if treeish in repo.refs.keys(base=b"refs/heads/"):
```
Comment (jelmer) [inline dulwich/objectspec.py:57]: perhaps just call parse_ref() here, so it also e.g. handles tags?

File: dulwich/porcelain.py (hunk 16 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1836]: Have  I update index in the stage()?Or I should update index on more time?

File: dulwich/porcelain.py (hunk 17 above)
Comment (jelmer) [reply dulwich/porcelain.py:1759]: presumably if you delete the space before rather than after "str"?

File: dulwich/porcelain.py (hunk 16 above)
Comment (jelmer) [reply dulwich/porcelain.py:1836]: So ideally we'd just update the index once for each checkout operation, since it can be expensive for larger indexes. I'm okay with compromising for now and e.g. doing one write for staging and one for unstaging, but one write for each will be noticeable.

Comment (DedSecer) [reply dulwich/porcelain.py:1836]: sorry, I can’t understand you.Since we have update index in the Repo.unstage, should we update index in the end of checkout?

File: dulwich/porcelain.py (hunk 18 above)
Comment (DedSecer) [reply dulwich/porcelain.py]: using tree.lookup_path to check if the file in the tree now.

File: dulwich/objectspec.py (hunk 19 above)
Comment (DedSecer) [reply dulwich/objectspec.py]: using parse_ref() now

File: dulwich/porcelain.py (hunk 16 above)
Comment (jelmer) [reply dulwich/porcelain.py:1847]: Yeah, we should update the index at the end of checkout - but we can't be updating the index for every reset_file() call, since that simply won't scale.

Comment (DedSecer) [reply dulwich/porcelain.py:1847]: Repo.unstage will update the different file's index between tatget branch and current branch, should we update the index at the end of checkout.

Comment (jelmer) [reply dulwich/porcelain.py:1847]: Yes -  we shouldn't be modifying the index every time reset_file() gets called. The changes *after* unstage still need to be properly written to the index.

Comment (DedSecer) [reply dulwich/porcelain.py:1847]: So I think we just update index after unstage and needn't update index at the end of checkout.

**Comment by jelmer** [comment]
Sorry, I've been travelling. I'll take a look at this tomorrow or on Tuesday.

File: dulwich/objects.py [hunk 20]
```diff
@@ -1179,7 +1179,12 @@ def lookup_path(self, lookup_obj, path):
           path: Path to lookup
         Returns: A tuple of (mode, SHA) of the resulting path.
         """
-        parts = path.split(b"/")
```
Comment (jelmer) [inline dulwich/objects.py:1182]: Why is this necessary? At the very least, there should be accompanying test changes - but I don't understand what bug in lookup_path this is fixing.

File: dulwich/objectspec.py [hunk 21]
```diff
@@ -54,6 +54,8 @@ def parse_tree(repo, treeish):
       KeyError: If the object can not be found
     """
     treeish = to_bytes(treeish)
+    if treeish not in repo.object_store.__iter__():
```
Comment (jelmer) [inline dulwich/objectspec.py:57]: Please don't use __iter__ but __contains__ - __iter__ can be very slow on large object stores. Why is the code on line 59 not sufficient? Tha twould have the same ffect.

File: dulwich/repo.py [hunk 22]
```diff
@@ -1305,6 +1305,58 @@ def stage(self, fs_paths):
...
+
+        from dulwich.index import IndexEntry
+        from dulwich.porcelain import remove
+
+        index = self.open_index()
+        try:
+            tree_id = self[b'HEAD'].tree
+        # no head mean no commit in the repo
+        except KeyError:
+            remove(self, paths=[os.path.join(self.path, path.decode()) for path in relpaths], cached=True)
```
Comment (jelmer) [inline dulwich/repo.py:1323]: porcelain is low level code, please don't add circular imports from it into dulwich.repo. It should be the other way around - dulwich.porcelain.remove() calls Repo.unstage.

File: dulwich/objects.py (hunk 20 above)
Comment (DedSecer) [reply dulwich/objects.py:1182]: this [comment](https://github.com/dulwich/dulwich/pull/898#discussion_r692582168)

File: dulwich/objectspec.py (hunk 21 above)
Comment (DedSecer) [reply dulwich/objectspec.py:57]: I would like to match treeish in the `repo.refs` while treeish is not a **commit sha** . It seem it isn't a good idea. Do you have better solution? what about replace `repo.object_store.__iter__()` with `list(repo.object_store)`?

File: dulwich/repo.py (hunk 22 above)
Comment (DedSecer) [reply dulwich/repo.py]: updated in the new commit

File: dulwich/porcelain.py [hunk 23]
```diff
@@ -1753,6 +1755,89 @@ def update_head(repo, target, detached=False, new_branch=None):
...
+        for file in get_tree_changes(repo)['modify']:
+            if file in repo.open_index():
+                raise DirNotCleanError('trying to checkout when working directory not clean')
+        
+        index = repo.open_index()
+        normalizer = repo.get_blob_normalizer()
+        filter_callback = normalizer.checkin_normalize
+        unstaged_changes = list(get_unstaged_changes(index, repo.path, filter_callback))
+        for file in unstaged_changes:
+            if file in repo.open_index():
```
[the remark this answers is not in the document]
Comment (DedSecer) [reply dulwich/porcelain.py]: updated in the new commit

File: dulwich/objects.py (hunk 20 above)
Comment (jelmer) [reply dulwich/objects.py:1182]: You're passing a file system path to a function that expects a canonical git path - the caller there should be converting from a filesystem path to a git path and then calling lookup_path().

File: dulwich/objectspec.py (hunk 21 above)
Comment (jelmer) [reply dulwich/objectspec.py:57]: You should be able to rely on \_\_contains\_\_, or better yet - rely on just accessing the treeish and handling the exception if that doesn't work.

\_\_iter\_\_ and list() both traverse all the objects in the object store, which can be millions of objects for larger projects like the linux kernel.

File: dulwich/porcelain.py (hunk 16 above)
Comment (jelmer) [reply dulwich/porcelain.py:1831]: You'd have to pass index to reset_file() in that case though, and you may want to make it private - or provide a _reset_file() that works on an index and is used by both checkout() and a public reset_file() wrapper that takes a repo argument.

**Review by jelmer — COMMENTED** [review]
Another round of review comments. I'll see if I can cherry-pick at least the unstage changes.

File: dulwich/objects.py (hunk 20 above)
Comment (jelmer) [reply dulwich/objects.py:1182]: E.g. by calling dulwich.index._fs_to_tree_path,

File: dulwich/objectspec.py (hunk 21 above)
Comment (DedSecer) [reply dulwich/objectspec.py]: Thanks, updated in the new commit

File: dulwich/objects.py (hunk 20 above)
Comment (DedSecer) [reply dulwich/objects.py]: call dulwich.index._fs_to_tree_path now

File: dulwich/porcelain.py (hunk 16 above)
Comment (DedSecer) [reply dulwich/porcelain.py:1833]: why should I pass index to reset_file()? I needn't index in reseet_file().

**Comment by jelmer** [comment]
I've merged some of the components.

**Comment by DedSecer** [comment]
@jelmer Hey,are you still merging my pull requests into dulwich?

**Comment by jelmer** [comment]
Sorry for the delay - I'll take another look this week.

**Comment by jelmer** [comment]
Closing this now that #1148 has landed.
